vhdx/src/vhdx_metadata.rs

```diff
@@ -0,0 +1,285 @@
+                    let mut disk_id_slice = [0u8; 16];
+                    f.read_exact(&mut disk_id_slice)
+                        .map_err(Error::ReadMetadata)?;
+                    disk_spec.disk_id = LittleEndian::read_u128(&disk_id_slice[0..16]);
+                    metadata_presence |= METADATA_VIRTUAL_DISK_ID_PRESENT;
+                }
+
+                METADATA_LOGICAL_SECTOR_SIZE => {
+                    let mut logical_sector_size_slice = [0u8; 4];
+                    f.read_exact(&mut logical_sector_size_slice)
+                        .map_err(Error::ReadMetadata)?;
+                    disk_spec.logical_sector_size =
+                        LittleEndian::read_u32(&logical_sector_size_slice[0..4]);
+
+                    if !(disk_spec.logical_sector_size == 512
+                        || disk_spec.logical_sector_size == 4096)
+                    {
+                        return Err(Error::InvalidLogicalSectorSize);
+                    }
+                    metadata_presence |= METADATA_LOGICAL_SECTOR_SIZE_PRESENT;
+                }
+
+                METADATA_PHYSICAL_SECTOR_SIZE => {
+                    let mut physical_sector_size_slice = [0u8; 4];
+                    f.read_exact(&mut physical_sector_size_slice)
+                        .map_err(Error::ReadMetadata)?;
+                    disk_spec.physical_sector_size =
+                        LittleEndian::read_u32(&physical_sector_size_slice[0..4]);
+
+                    if !(disk_spec.physical_sector_size == 512
+                        || disk_spec.physical_sector_size == 4096)
+                    {
+                        return Err(Error::InvalidPhysicalSectorSize);
+                    }
+                    metadata_presence |= METADATA_PHYSICAL_SECTOR_SIZE_PRESENT;
+                }
+
+                METADATA_PARENT_LOCATOR => {
+                    metadata_presence |= METADATA_PARENT_LOCATOR_PRESENT;
+                }
+
+                _ => return Err(Error::InvalidMetadataItem),
+            }
+
+            // Check if all required metadata are present
+            if metadata_presence != METADATA_ALL_PRESENT {
+                return Err(Error::MissingMetadata);
+            }
+
+            if (metadata_entry.flag_bits & METADATA_FLAGS_IS_REQUIRED) != 0 {
+                return Err(Error::UnsupportedFlag);
+            }
+
+            // Check if the virtual disk size is a multiple of the logical sector size
+            if ((metadata_presence & METADATA_LOGICAL_SECTOR_SIZE_PRESENT) != 0)
+                && (disk_spec.virtual_disk_size % disk_spec.logical_sector_size as u64 != 0)
+            {
+                return Err(Error::InvalidBlockSize);
+            }
+
+            disk_spec.sectors_per_block = disk_spec.block_size / disk_spec.logical_sector_size;
+            disk_spec.chunk_ratio = (MAX_SECTORS_PER_BLOCK * disk_spec.logical_sector_size
+                / disk_spec.block_size) as u64;
```

Comment:
For these two, it might be worth writing two very small functions. This would increase readability.

Reply:
Done.